luma/lcd/device.py

```diff
@@ -757,3 +760,234 @@ def contrast(self, value):
         """
         assert(0 <= value <= 255)
         self.command(0x81, value >> 2)
+
+
+class hd44780(parallel_device, character):
+    """
+    Driver for a HD44780 style LCD display.  This class provides a ``text``
+    property which can be used to set and get a text value, which will be
+    rendered to the display's screen using the display's built-in font.
+
+    :param serial_interface: The serial interface (usually a
+        :py:class:`luma.core.interface.serial.parallel` instance) to delegate
+        sending data and commands through.
+    :param width: The number of characters that can be displayed on a line.
+    :type width: int
+    :param height: The number of lines the display supports.
+    :type height: int
+    :param undefined: character to use if a requested character is not in the
+        font tables
+    :type undefined: str
+    :param selected_font: the font table appropriate for the model of display you
+        are using.  The hd44780 normally comes in a version with font A00 (ENGLISH_JAPANESE) or A02 (ENGLISH_EUROPEAN).  You can provide either
+        the name ('A00' or 'A02') or the number (0 for 'A00', 1 for 'A02') for
+        the font your display contains.
+    :type selected_font: int or str
+    :param exec_time: Time in seconds to wait for a command to complete.
+        Default is 50 μs (1e-6 * 50) which typically is long enough for commands
+        to finish.  If your display is not working correctly, you may want to
+        try increasing the exec_time delay.
+    :type exec_time: float
+    :param framebuffer: Framebuffering strategy, currently values of
+        ``diff_to_previous`` or ``full_frame`` are only supported.
+    :type framebuffer: str
+
+    To place text on the display, simply assign the text to the 'text'
+    instance variable::
+
+        p = parallel(RS=7, E=8, PINS=[25,24,23,18])
+        my_display = hd44780(p, selected_font='A00')
+        my_display.text = 'HD44780 Display\\nFont A00 Eng/Jap'
+
+    For more details on how to use the 'text' interface see
+    :class:`luma.core.virtual.character`
+
+    ..note:
+        This driver currently only supports the hd44780 5x8 display mode.
+
+    .. versionadded:: 2.5.0
+    """
+    def __init__(self, serial_interface=None, width=16, height=2, undefined='_',
+                 selected_font=0, exec_time=1e-6 * 50, framebuffer="diff_to_previous", **kwargs):
```

Comment:
value for exec_time is a constant in luma.core now, can you use that here?

Reply:
Oh PULSE_TIME isn't part of `__all__` though. Maybe it should be moved to `__all__` or a different place so it can be re-used in sub-projects.

Reply:
Actually pulse_time and exec_time are different values.  pulse_time sets the duration of the 'enable' pulse (low-high-low) on the enable pin of a parallel-bus display.  Exec_time is to give the display time to process a command before it receives another message.  

In recent testing, I've been examining the lower-limit for using sleep to create these delays.  At least on RPis using Raspbian, it appears that the smallest delay you can request is 1.2e-5.  So the 1e-6 isn't really giving me the correct timing anyway.  Eventually, I'll look into seeing if there is another method that I can use to get closer to the spec sheets for the displays but for now, slower still works.

As far as correct placement in the code, each parallel device has its own requirements but 1e-5 is likely to be above the required value for pretty much any display.  So, leaving it in luma.core.interface.serial makes sense.

Reply:
sounds good.

Reply:
I went down a rabbit hole trying to get better GPIO performance by writing a c++ class using libgpiod.  Bottom line is even at the speeds I could get out of that library, the required pulse_time for the ws0010 is shorter than the time libgpiod takes to move the pin from low to high to low.  So effectively, the default pulse_time for the ws0010 (and winstar_weh) can be set to 0.  This significantly improves the update time for that display.  However, if you are trying to horizontally scroll anything, the upper limit appears to be around 30 FPS.  Above that, animations show tearing.  

I haven't done as much testing of the hd44780 displays but they appear to be much slower devices.  Setting pulse_time to zero does cause stability problems so we are stuck using at least some delay.  For now, I'll just leave the default to 1e-6 but may want to change from using sleep to a timing loop in future versions.  Also, 5 FPS was the sweet spot for refreshing it.